tests/test_smoke.py

```diff
@@ -609,3 +611,34 @@ def test_msvc_x86(self, capfd, basic_cmake_project):
         run(f'cmake -S {source_dir} -B {binary_dir} -DCMAKE_PROJECT_TOP_LEVEL_INCLUDES={conan_provider} -G "Visual Studio 16 2019" -A Win32')
         out, _ = capfd.readouterr()
         assert "arch=x86" in out
+
+
+class TestCMakeDepsGenerators:
+    # CMakeDeps generator is declared in the generate() function in conanfile.py
+    def test_one_generator(self, capfd, basic_cmake_project):
+        source_dir, binary_dir = change_conanfile(basic_cmake_project, 'single_generator')
+        run(f'cmake -S {source_dir} -B {binary_dir} -DCMAKE_PROJECT_TOP_LEVEL_INCLUDES={conan_provider} -DCMAKE_BUILD_TYPE=Release')
+        out, _ = capfd.readouterr()
+        assert 'Generating done' in out
+
+    # CMakeDeps generator is declared both in generators attribute and generate() function in conanfile.py
+    def test_duplicate_generator(self, capfd, basic_cmake_project):
+        source_dir, binary_dir = change_conanfile(basic_cmake_project, 'duplicate_generator')
+        run(f'cmake -S {source_dir} -B {binary_dir} -DCMAKE_PROJECT_TOP_LEVEL_INCLUDES={conan_provider} -DCMAKE_BUILD_TYPE=Release', check=False)
+        _, err = capfd.readouterr()
+        assert ('ConanException: CMakeDeps is declared in the generators attribute, but was instantiated in the '
+                'generate() method too') in err
+
+    # CMakeDeps generator is not declared in either place
+    def test_no_generator(self, capfd, basic_cmake_project):
+        source_dir, binary_dir = change_conanfile(basic_cmake_project, 'no_generator')
+        run(f'cmake -S {source_dir} -B {binary_dir} -DCMAKE_PROJECT_TOP_LEVEL_INCLUDES={conan_provider} -DCMAKE_BUILD_TYPE=Release', check=False)
+        out, _ = capfd.readouterr()
+        assert 'Could NOT find hello (missing: hello_DIR)' in out
+
+def change_conanfile(basic_cmake_project, gen_resource):
+    source_dir, binary_dir = basic_cmake_project
+    os.chdir(source_dir.as_posix())
```

Comment:
the `basic_cmake_project` fixture will be invoked when entering the test, which already does a `chdir` to the build directory for the duration of the test.
This line is changing the cwd to the source directory (so, overriding the fixture's intention), but also in a way that is persisted across the session (there's no yield or anything restoring it back to the original value). 

`source_dir` should already be a full path in the call to `shutil.copy2` so perhaps unnecessary to do a chdir